jax/experimental/mosaic/gpu/examples/fdgg.py

```diff
@@ -0,0 +1,505 @@
+  def __repr__(self):
+    x = self.x
+    return f"Warpgroup({x=}, range={x*WARPSIZE}-{(x+4)*WARPSIZE-1})"
+
+
+@dataclass
+class WarpAllocator:
+  def __init__(self):
+    self.warp_free = MAX_WARPS_PER_BLOCK_DIM * [True]
+
+  def alloc_warp(self):
+    i = self.warp_free.index(True)
+    result = Warp(i)
+    self.warp_free[i] = False
+    return result
+
+  def alloc_warpgroup(self):
+    i = None
+    for j in range(0, len(self.warp_free), 4):
+      if all(self.warp_free[j:j+4]):
+        i = j
+        break
+    assert i is not None
+    result = Warpgroup(i)
+    for j in range(4):
+      self.warp_free[i+j] = False
+    return result
+
+  def block_size(self):
+    return WARPSIZE * (len(self.warp_free) - self.warp_free[::-1].index(False))
+
+
+@contextlib.contextmanager
+def _only(w, warpid):
+  index = ir.IndexType.get()
+  if isinstance(w, Warp):
+    is_warp_x = arith.cmpi(arith.CmpIPredicate.eq, warpid, c(w.x,index))
+    with ir.InsertionPoint(scf.IfOp(is_warp_x).then_block):
+      yield
+      scf.yield_([])
+  elif isinstance(w, Warpgroup):
+    a = w.x
+    b = a+4
+    gea = arith.cmpi(arith.CmpIPredicate.uge, warpid, c(a,index))
+    ltb = arith.cmpi(arith.CmpIPredicate.ult, warpid, c(b,index))
+    predicate = arith.andi(gea, ltb)
+    with ir.InsertionPoint(scf.IfOp(predicate).then_block):
+      yield
+      scf.yield_([])
+  else:
+    raise ValueError
+
+
+@contextlib.contextmanager
+def single_warp_thread():
+  i1 = ir.IntegerType.get_signless(1)
+  elected = nvvm.elect_sync(i1)
+  with ir.InsertionPoint(scf.IfOp(elected).then_block):
+    yield
+    scf.yield_([])
+
+
+def build_kernel(
+    m, n, k,
+    cta_count: int,
+    expert_count: int,
+    tile_m: int = 128,
+    tile_n: int = 128,
+    max_concurrent_steps: int = 2,
+    in_dtype = jnp.float16,
+    profiler_spec: profiler.ProfilerSpec | None = None,
+):
+  i1 = ir.IntegerType.get_signless(1)
+  i32 = ir.IntegerType.get_signless(32)
+  index = ir.IndexType.get()
+  f32 = ir.F32Type.get()
+
+  cx = lambda v: c(v, index)
+
+  swizzle = 128
+  swizzle_elems = tile_k = swizzle // 2
+  tiling = (8, swizzle_elems)
+  tmem_slot_count = 2
+
+  k_loop_iter = k // tile_k
+  max_concurrent_steps = min(max_concurrent_steps, k_loop_iter)
+
+  if n % tile_n != 0:
+    raise ValueError(f"{n=} must be divisible by {tile_n=}")
+  if k % tile_k != 0:
+    raise ValueError(f"{k=} must be divisible by {tile_k=}")
+  if n % expert_count != 0:
+    raise ValueError(f"{n=} must be divisible by {expert_count=}")
+
+  expert_n = n // expert_count
+
+  if expert_n % tile_n != 0:
+    raise ValueError(f"{expert_n=} must be divisible by {tile_n=}")
+
+  n_tile_count = expert_n // tile_n
+
+  def ceildiv(x, y):
+    return arith.divui(arith.subi(arith.addi(x, y),
+                                  cx(1)),
+                       y)
+
+  allocator = WarpAllocator()
+  warp = allocator.alloc_warp
+  warpgroup = allocator.alloc_warpgroup
+
+  tma_warp = warp()
+  mma_warp = warp()
```

Comment:
What's the benefit of the whole WarpAllocator indirection? They're still allocated statically, so you can just hardcode them to `TMA_WARP = 0` and `MMA_WARP = 1` as you did before, right?